[inductor][BE] don't try/except ImportError for AttrsDescriptor versions

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144807

motivation: Ed's advice to avoid `except ImportError` (i.e. based on the fact that your target module/class might in fact exist, but you might run into some different ImportError whose stacktrace you now ignore).

additional motivation: I'm going to add some more cases to this list, and would like to avoid this pattern:
```
try:
   ...
except ImportError:
    try:
        ...
    except ImportError:
        try:
            ...
```

suggestions on better ways to do this would be appreciated!

test: ran with triton commit e5be006a (last working commit) and 34a6a2ff8 (in june, when AttrsDescriptor was still in triton.compiler.compiler)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov @BoyuanFeng

davidberard98 committed:
[inductor][BE] don't try/except ImportError for AttrsDescriptor versions

motivation: Ed's advice to avoid `except ImportError` (i.e. based on the fact that your target module/class might in fact exist, but you might run into some different ImportError whose stacktrace you now ignore).

additional motivation: I'm going to add some more cases to this list, and would like to avoid this pattern:
```
try:
   ...
except ImportError:
    try:
        ...
    except ImportError:
        try:
            ...
```

suggestions on better ways to do this would be appreciated!

test: ran with triton commit e5be006a (last working commit) and 34a6a2ff (in june, when AttrsDescriptor was still in triton.compiler.compiler)

[ghstack-poisoned]

davidberard98 committed:
Update on "[inductor][BE] don't try/except ImportError for AttrsDescriptor versions"


motivation: Ed's advice to avoid `except ImportError` (i.e. based on the fact that your target module/class might in fact exist, but you might run into some different ImportError whose stacktrace you now ignore).

additional motivation: I'm going to add some more cases to this list, and would like to avoid this pattern:
```
try:
   ...
except ImportError:
    try:
        ...
    except ImportError:
        try:
            ...
```

suggestions on better ways to do this would be appreciated!

test: ran with triton commit e5be006a (last working commit) and 34a6a2ff8 (in june, when AttrsDescriptor was still in triton.compiler.compiler)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov BoyuanFeng

[ghstack-poisoned]

davidberard98 committed:
Update on "[inductor][BE] don't try/except ImportError for AttrsDescriptor versions"


motivation: Ed's advice to avoid `except ImportError` (i.e. based on the fact that your target module/class might in fact exist, but you might run into some different ImportError whose stacktrace you now ignore).

additional motivation: I'm going to add some more cases to this list, and would like to avoid this pattern:
```
try:
   ...
except ImportError:
    try:
        ...
    except ImportError:
        try:
            ...
```

suggestions on better ways to do this would be appreciated!

test: ran with triton commit e5be006a (last working commit) and 34a6a2ff8 (in june, when AttrsDescriptor was still in triton.compiler.compiler)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov BoyuanFeng

[ghstack-poisoned]

c00w commented on torch/_inductor/runtime/hints.py:
```diff
@@ -64,7 +67,7 @@ def AttrsDescriptorWrapper(
             return res
 
-    except ImportError:
+    elif hasattr(triton.compiler.compiler, "AttrsDescriptor"):
```
Correctness - Should this always be defined? (and if not we should throw?). Otherwise if triton is available, but somehow we're missing AttrsDescriptor on triton.compiler.compiler, you'll end up with import errors further down the stack that are less descriptive.

davidberard98 replied:
Good point, I'll make this `else` instead of `elif`.

And whether `triton.compiler.compiler.AttrsDescriptor` is always defined - it was removed on Jan 1. So my follow-up PR is going to be to add handling for the new semantics.

ezyang reviewed (approved):
Colin's comment is valid but I am not sure we can distinguish sufficiently old Triton and too new Triton. Maybe the old Triton is just not supported anymore so the else branch can hard error.

Skylion007 commented:
Would it better use importlib here for more control over the importing / handling if a module exists or not without actually importing it?

davidberard98 commented:
@Skylion007 do you have a recommendation for how to do this with`importlib`? Ed's suggestion was to use `importlib.util.find_spec` - which (AFAICT) works for finding modules (and whether or not they exist), but not for checking whether a class exists on a module. In this case, I'm assuming the existence of these modules (triton.backends.compiler and triton.compiler.compiler) and just need to check where this class is defined.

Skylion007 commented:
Yeah, I don't think you can check if a class exists without importing
because it needs to run the module script to figure out what classes are
generated (for instance if multiple classes are dynamically generated)

On Wed, Jan 15, 2025 at 1:52 PM David Berard ***@***.***>
wrote:

> @Skylion007 <https://github.com/Skylion007> do you have a recommendation
> for how to do this withimportlib? Ed's suggestion was to use
> importlib.util.find_spec - which (AFAICT) works for finding modules (and
> whether or not they exist), but not for checking whether a class exists on
> a module. In this case, I'm assuming the existence of these modules
> (triton.backends.compiler and triton.compiler.compiler) and just need to
> check where this class is defined.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/pytorch/pytorch/pull/144807#issuecomment-2593705587>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPVMX4C6G37XHX2CYVDM6T2K2U6BAVCNFSM6AAAAABVF5R4JCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKOJTG4YDKNJYG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

davidberard98 committed:
Update on "[inductor][BE] don't try/except ImportError for AttrsDescriptor versions"


motivation: Ed's advice to avoid `except ImportError` (i.e. based on the fact that your target module/class might in fact exist, but you might run into some different ImportError whose stacktrace you now ignore).

additional motivation: I'm going to add some more cases to this list, and would like to avoid this pattern:
```
try:
   ...
except ImportError:
    try:
        ...
    except ImportError:
        try:
            ...
```

suggestions on better ways to do this would be appreciated!

test: ran with triton commit e5be006a (last working commit) and 34a6a2ff8 (in june, when AttrsDescriptor was still in triton.compiler.compiler)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang aakhundov BoyuanFeng

[ghstack-poisoned]

davidberard98 commented:
@pytorchbot merge
